Add Firecracker shutdown driver and some minor Kconfig fixups

mogasergiu opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.


### Base target

 - Architecture(s): `x86_64`
 - Platform(s):  `kvm`
 - Application(s): N/A


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

Fix some minor misindented lines in `lib/ukboot`'s KConfig and some minor mistakes in `drivers/`'s `Config.uk` files and implement a very basic dumbed down PS/2 keyboard driver, enough to be able to process the `SendCtrlAltDelete` request of Firecracker.

mogasergiu committed:
lib/ukboot: Fix `KConfig` styling of `Config.uk`

Maintain consistency among `KConfig` files and align `lib/ukboot`'s
`Config.uk` file's indentation with the others'.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
drivers: Add `PS/2` subsystem with a dumbed down PS2 keyboard driver

Add a directory to represent what will be in the future the
subsystem for PS/2 controller drivers subsystem.

Implement a basic, dumbed down, driver stub for the PS/2 keyboard.
Write just enough functionality to register an IRQ handler for the
well known default legacy PIC routed PS/2 keyboard IRQ line and be
able to process 3 scan codes received in a burst manner:
CTRL + ALT + DEL

This functionality is bare minimum required to achieve shutdown with
Firecracker's `SendCtrlAltDel` command. It may not work on QEMU and
is obviously not the way the real driver should be implemented.

Additionally, introduce a new invisible config option:
`CONFIG_HAVE_SHUTDOWN_DISPATCHER`
Now, whenever a component implements the functionality of raising a
shutdown event, it must announce it system-wide by selecting this
config. That being said, make the keyboard driver do this.
(In the future, when we will be able to receive actual keystrokes,
we may be able to make this a separate config of the keyboard driver,
besides the actual functionality of processing keystrokes).

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

michpappas reviewed (changes requested):
I think this should be under drivers, is there any specific reason it's being placed under `libukboot`?

mogasergiu commented:
> I think this should be under drivers, is there any specific reason it's being placed under `libukboot`?

Unfortunately this was a confusing decision for me to make as well... I did not place it into drivers because there would be no category for it or any exported interface, it's not a real driver but a dummy stub registering an IRQ handler for shutdown. As you can see I even registered it as a lib initcall... because we have no way of probing it, i.e. ACPI or ISAPNP. So I placed it here because `lib/ukboot` is the source of shutdown related stuff. I also renamed it dumbkbd because this is what linux names it when the PS2 controller is configured during probe to not wait a few milliseconds for some initialization stuff and it is also the argument passed by Firecracker to the commandline for this exact same reason.

mogasergiu commented on lib/ukboot/Makefile.uk:
```diff
@@ -13,6 +13,10 @@ endif
 LIBUKBOOT_SRCS-$(CONFIG_LIBUKBOOT_MAINTHREAD) += $(LIBUKBOOT_BASE)/shutdown_req.c|isr
 
+ifeq ($(CONFIG_KVM_VMM_FIRECRACKER),y)
+LIBUKBOOT_SRCS-$(CONFIG_LIBUKBOOT_MAINTHREAD) += $(LIBUKBOOT_BASE)/dumbkbd.c
```
Note for me: add `|isr`

michpappas commented:
> > I think this should be under drivers, is there any specific reason it's being placed under `libukboot`?
> 
> Unfortunately this was a confusing decision for me to make as well... I did not place it into drivers because there would be no category for it or any exported interface, it's not a real driver but a dummy stub registering an IRQ handler for shutdown. As you can see I even registered it as a lib initcall... because we have no way of probing it, i.e. ACPI or ISAPNP. So I placed it here because `lib/ukboot` is the source of shutdown related stuff. I also renamed it dumbkbd because this is what linux names it when the PS2 controller is configured during probe to not wait a few milliseconds for some initialization stuff and it is also the argument passed by Firecracker to the commandline for this exact same reason.

Actually it is not a hard requirement that drivers must implement an interface defined in `lib/`. Moreover it is not necessary to block until we introduce a subsystem. What I am trying to do here is avoid new PRs creating more work for re-arch, as every little new thing contributes to what eventually becomes bloat. So since this driver is started via init, is there a reason can't be moved to a new subsystem under `drivers/`?

mogasergiu committed:
drivers/ukintctlr: Delete unnecessary `exportsyms.uk`

The philosophy is that each driver is meant to individually export
whatever symbols it sees fit. Therefore, delete the unnecessary
`exportsyms.uk` in the root of the interrupt controller drivers
subsystem.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

michpappas commented on drivers/ukps2/dumbkbd.c:
```diff
@@ -0,0 +1,137 @@
+	while (!(inb(I8042_STATUS_REG) & I8042_STATUS_RECV_FULL)) {
+		if (unlikely(counter >= 5)) {
+			outb(I8042_CMD_REG, I8042_CMD_DIS_KBD);
+			uk_pr_err("PS/2 Controller unresponsie\n");
```
Should be "unresponsive".

michpappas commented on drivers/ukps2/dumbkbd.c:
```diff
@@ -0,0 +1,137 @@
+			return -ENODEV;
+		}
+
+		/* We are advised to wait 50ms if the controller hasn't
```
It would be useful to provide the origin of this advise.

michpappas commented on plat/kvm/Config.uk:
```diff
@@ -102,6 +102,7 @@ config KVM_VMM_FIRECRACKER
 	select KVM_BOOT_PROTO_LXBOOT
 	select HAVE_MMIO
 	select VIRTIO_MMIO_LINUX_COMPAT_CMDLINE
+	select LIBUKPM_PS2 if ARCH_X86_64 && LIBUKBOOT_MAINTHREAD
```
Should be `select LIBUKPS2`. Also `drivers/ukps2` needs a `Config.uk`.

Moreover the way to express the dependency between libukboot and this feature would be to define a `HAVE_PS2` in `drivers/ukps2/` as:
```
config HAVE_PS2
    bool

config LIBUKPS2
    bool "Minimal PS/2 driver for CtrlAltDel"
    depends on HAVE_PS2
```

Then in `KVM_VMM_FIRECRACKER`:
```
select HAVE_PS2
```

And finally in `libukboot`:
```
imply LIBUKBOOT_MAINTHREAD if HAVE_PS2
```

mogasergiu replied:
Following an offline discussion, it's been decided that we proceed as such:
- in plat/kvm/Config.uk for FC
select HAVEPS2
imply LIBUKPS2_KBD
- in lib/ukboot/Config.uk
imply LIBUKBOOT_MAINTHREAD if HAVE_SHUTDOWN
- in drivers/ukps2/kbd/Config.uk
select HAVE_SHUTDOWN # hey by selecting the kbd you now have shutdown available
- in plat/Config.uk
config HAVE_SHUTDOWN
- in lib/ukps2/Config.uk
config HAVE_PS2

michpappas commented on plat/kvm/Config.uk:
```diff
@@ -101,6 +101,8 @@ config KVM_VMM_FIRECRACKER
 	bool "Firecracker"
 	select KVM_BOOT_PROTO_LXBOOT
 	select HAVE_MMIO
+	select HAVE_PS2
```
You also need to `select HAVE_SHUTDOWN` here. In the current state, `UKBOOT_MAINTHREAD` is not selected.

mogasergiu replied:
I will make it so that the keyboard driver selects it isntead, as  if announcing "hey, enabling me registers shutdown handler"

michpappas replied:
Sounds good. Besides, the result is the same: shutdown is still enabled but explicitly. One thing is that in that case we assume that Unikraft will always be able to shutdown, as long as a method to initiate shutdown is provided to the user. If the platform does not support it, then it will probably ignore the event, but Unikraft will go all the way to halt nevertheless. One thing we could improve then is renaming `HAVE_SHUTDOWN` to something that indicates more clearly that we are referring to having a shutdown method, to avoid confusion.

Also if possible, please update the commit message with a brief description of the design.

michpappas commented on plat/Config.uk:
```diff
@@ -119,3 +119,8 @@ config FPSIMD
+
+config HAVE_SHUTDOWN
+	bool
+	help
```
This item is invisible, so `help` is redundant.

michpappas commented on drivers/ukps2/kbd/Config.uk:
```diff
@@ -0,0 +1,3 @@
+config LIBUKPS2_KBD
+	bool "PS/2 Keyboard dumbed down driver"
+	depends on ARCH_X86_64
```
This needs `depends on HAVE_PS2` so that it's not visible on architectures that don't have that capability.

michpappas commented on lib/ukboot/Config.uk:
```diff
@@ -4,6 +4,7 @@ menuconfig LIBUKBOOT
 	select LIBUKDEBUG
 	select LIBUKARGPARSE
 	select HAVE_BOOTENTRY
+	imply LIBUKBOOT_MAINTHREAD if HAVE_SHUTDOWN
```
You also need to add `imply LIBUKBOOT_SHUTDOWNREQ_HANDLER if HAVE_SHUTDOWN`.

michpappas commented on plat/kvm/Config.uk:
```diff
@@ -101,6 +101,8 @@ config KVM_VMM_FIRECRACKER
 	select KVM_BOOT_PROTO_LXBOOT
 	select HAVE_MMIO
+	select HAVE_PS2
+	imply LIBUKPS2_KBD
```
Although the PS2 driver already depends on `ARCH_X86_64`, I find it looks like it's universally enabled here, so perhaps change to `select HAVE_PS2 if ARCH_X86_64` and `imply LIBUKPS2 if ARCH_X86_64`? We also follow that pattern for GIC and xPIC.

michpappas reviewed (changes requested):
@mogasergiu a few comments on the Kconfig update.

michpappas reviewed (approved):
@mogasergiu thanks for the updates, I think it looks good now 🚀 

Reviewed-by: Michalis Pappas <michalis@unikraft.io>

michpappas commented on drivers/Config.uk:
```diff
@@ -27,3 +27,9 @@ config HAVE_IBMPC
 	select HAVE_IBMPC_NS16550
 	select HAVE_IBMPC_VGA
+
+menu "PS/2 controller"
```
Please move this up so that config options appear in alphabetical order:
```diff
diff --git a/drivers/Config.uk b/drivers/Config.uk
index 497c02304..2a39fa370 100644
--- a/drivers/Config.uk
+++ b/drivers/Config.uk
@@ -6,6 +6,10 @@ menu "Interrupt controller"
 source "$(shell,$(UK_BASE)/support/build/config-submenu.sh -q -o '$(KCONFIG_DIR)/drivers-intctlr.uk' -r '$(KCONFIG_DRIV_BASE)/ukintctlr' -l '$(KCONFIG_DRIV_BASE)/ukintctlr' -e '$(KCONFIG_EX
CLUDEDIRS)')"
 endmenu
 
+menu "PS/2 controller"
+source "$(shell,$(UK_BASE)/support/build/config-submenu.sh -q -o '$(KCONFIG_DIR)/drivers-ukps2.uk' -r '$(KCONFIG_DRIV_BASE)/ukps2' -l '$(KCONFIG_DRIV_BASE)/ukps2' -e '$(KCONFIG_EXCLUDEDIRS)
')"
+endmenu
+
 menu "Random Number Generator"
 source "$(shell,$(UK_BASE)/support/build/config-submenu.sh -q -o '$(KCONFIG_DIR)/drivers-random.uk' -r '$(KCONFIG_DRIV_BASE)/ukrandom' -l '$(KCONFIG_DRIV_BASE)/ukrandom' -e '$(KCONFIG_EXCLU
DEDIRS)')"
 endmenu
@@ -26,10 +30,5 @@ endmenu
 config HAVE_IBMPC
        bool
        select HAVE_IBMPC_NS16550
+       select HAVE_IBMPC_PS2
        select HAVE_IBMPC_VGA
-
-menu "PS/2 controller"
-
-source "$(shell,$(UK_BASE)/support/build/config-submenu.sh -q -o '$(KCONFIG_DIR)/drivers-ukps2.uk' -r '$(KCONFIG_DRIV_BASE)/ukps2' -l '$(KCONFIG_DRIV_BASE)/ukps2' -e '$(KCONFIG_EXCLUDEDIRS)
')"
-
-endmenu
```
**PS:** For the `HAVE_IBMPC_PS2` see the comments below.

michpappas commented on drivers/Makefile.uk:
```diff
@@ -13,3 +13,4 @@ $(eval $(call import_lib,$(UK_DRIV_BASE)/ukrandom))
 $(eval $(call import_lib,$(UK_DRIV_BASE)/ukrtc))
 $(eval $(call import_lib,$(UK_DRIV_BASE)/virtio))
 $(eval $(call import_lib,$(UK_DRIV_BASE)/xen))
+$(eval $(call import_lib,$(UK_DRIV_BASE)/ukps2))
```
Same about alphabetical order here:
```diff
diff --git a/drivers/Makefile.uk b/drivers/Makefile.uk
index 7431dfcd4..96099a7e8 100644
--- a/drivers/Makefile.uk
+++ b/drivers/Makefile.uk
@@ -9,8 +9,8 @@ UK_DRIV_BASE := $(CONFIG_UK_BASE)/drivers
 $(eval $(call import_lib,$(UK_DRIV_BASE)/ukbus))
 $(eval $(call import_lib,$(UK_DRIV_BASE)/ukconsole))
 $(eval $(call import_lib,$(UK_DRIV_BASE)/ukintctlr))
+$(eval $(call import_lib,$(UK_DRIV_BASE)/ukps2))
 $(eval $(call import_lib,$(UK_DRIV_BASE)/ukrandom))
 $(eval $(call import_lib,$(UK_DRIV_BASE)/ukrtc))
 $(eval $(call import_lib,$(UK_DRIV_BASE)/virtio))
 $(eval $(call import_lib,$(UK_DRIV_BASE)/xen))
-$(eval $(call import_lib,$(UK_DRIV_BASE)/ukps2))
```

michpappas commented on drivers/ukps2/Config.uk:
```diff
@@ -0,0 +1,2 @@
+config HAVE_PS2
```
Now that we have `HAVE_IBMPC` you can make PS2 part of it. Also you can move the dependency of x86_64 here as it applies to all ps2 drivers.
```diff
diff --git a/drivers/ukps2/Config.uk b/drivers/ukps2/Config.uk
index e923767df..8815c6c63 100644
--- a/drivers/ukps2/Config.uk
+++ b/drivers/ukps2/Config.uk
@@ -1,2 +1,3 @@
-config HAVE_PS2
+config HAVE_IBMPC_PS2
        bool
+       depends on ARCH_X86_64
```

michpappas commented on drivers/ukps2/kbd/Config.uk:
```diff
@@ -0,0 +1,5 @@
+config LIBUKPS2_KBD
```
Ditto for `HAVE_IBMPC_PS2` and moving the arch dependency one level up:
```diff
diff --git a/drivers/ukps2/kbd/Config.uk b/drivers/ukps2/kbd/Config.uk
index 94b67025c..e1c190d05 100644
--- a/drivers/ukps2/kbd/Config.uk
+++ b/drivers/ukps2/kbd/Config.uk
@@ -1,5 +1,4 @@
 config LIBUKPS2_KBD
        bool "PS/2 Keyboard dumbed down driver"
-       depends on ARCH_X86_64
-       depends on HAVE_PS2
+       depends on HAVE_IBMPC_PS2
        select HAVE_SHUTDOWN_DISPATCHER
```

michpappas commented on plat/kvm/Config.uk:
```diff
@@ -116,6 +116,8 @@ config KVM_VMM_FIRECRACKER
 	imply LIBNS16550 if LIBUKCONSOLE
 	imply LIBNS16550_EARLY_CONSOLE if LIBUKCONSOLE
 	imply LIBNS16550_COM1 if LIBUKCONSOLE && ARCH_X86_64
+	select HAVE_PS2 if ARCH_X86_64
```
Besides renaming to `HAVE_IBMPC_PS2` you can also move this together with the `HAVE_IBMPC_PS2_NS16550` directive:
```diff
diff --git a/plat/kvm/Config.uk b/plat/kvm/Config.uk
index e47f76191..f95d795bd 100644
--- a/plat/kvm/Config.uk
+++ b/plat/kvm/Config.uk
@@ -110,13 +110,13 @@ config KVM_VMM_QEMU
 config KVM_VMM_FIRECRACKER
        bool "Firecracker"
        select KVM_BOOT_PROTO_LXBOOT
        select HAVE_IBMPC_NS16550 if ARCH_X86_64
+       select HAVE_IBMPC_PS2 if ARCH_X86_64
        select HAVE_MMIO
        select VIRTIO_MMIO_LINUX_COMPAT_CMDLINE if LIBVIRTIO_MMIO
        imply LIBNS16550 if LIBUKCONSOLE
        imply LIBNS16550_EARLY_CONSOLE if LIBUKCONSOLE
        imply LIBNS16550_COM1 if LIBUKCONSOLE && ARCH_X86_64
-       select HAVE_PS2 if ARCH_X86_64
        imply LIBUKPS2_KBD if ARCH_X86_64
        imply LIBUKINTCTLR_GICV3 if ARCH_ARM_64
        imply LIBUKRTC_PL031
```

michpappas reviewed (changes requested):
@mogasergiu only a few comments related to the addition of `HAVE_IBMPC`, after that we can merge.

mogasergiu commented:
Applied patches. Thank you so much for taking the time to provide diffs! :pray:
Saved me a lot of time!

michpappas reviewed (approved):
Reviewed-by: Michalis Pappas <michalis@unikraft.io>

michpappas commented:
Approved-by: Michalis Pappas <michalis@unikraft.io>
